source/MaterialXCore/Element.h

```diff
@@ -761,7 +761,12 @@ class Element : public std::enable_shared_from_this<Element>
     ///    geometry token substitutions are applied.  If the universal geometry
     ///    name "/" is given, then all geometry token substitutions are applied,
     /// @return A shared pointer to a StringResolver.
-    StringResolverPtr createStringResolver(const string& geom = EMPTY_STRING) const;
+    virtual StringResolverPtr createStringResolver(const string& geom = EMPTY_STRING) const;
+
+    /// Add tokens to string resolver. Derived classes can override this
+    /// method to add per-class type tokens as desired. The default is to not
+    /// add any tokens.
+    virtual void addTokens(StringResolverPtr& /*resolver*/) const;
```

Comment:
addTokens() is a new virtual to allow each class derived from Element to add tokens to a string resolve as desired.
I used Element since it's the only common base class for InterfaceElement (ports) and GeomInfo.